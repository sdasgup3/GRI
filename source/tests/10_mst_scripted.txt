function displayAdjacencyMatrix(matrix, vertices)
{
  print("   ");
  foreach(vertex; vertices.iterator()) {
    print("" + vertex.__id + " ");
  }

    println("");
  it = vertices.iterator();
  foreach(line; matrix.iterator()) {
    if(it.hasNext())
      vertex = it.next();

    print("" + vertex.__id + " |");
        
    foreach(item; line.iterator())  {
      print("" + item + " ");
    }

    println("");
  }
}

function mst_helper(g)
{
  vertices = g.getVertices();
  numvertices = g.getNumVertices();
  edges = g.getEdges();
  numedges = g.getNumEdges();

  //Init dist matrix
  matrix  = array(numvertices);
  for(i = 0; i < numvertices; i++) {
    matrix[i] = array(numvertices);
    for(j = 0 ; j < numvertices; j++) {
      matrix[i][j] = 0.0;
    }
  }

  foreach(edge; edges.iterator()) {
    begin = edge.getBeginVertex();
    end = edge.getEndVertex();
    wt = edge.__w;

    matrix[begin.__id][end.__id] = wt;
    matrix[end.__id][begin.__id] = wt;
  }

  displayAdjacencyMatrix(matrix, vertices);

  pathNdist = mst(matrix, numvertices);
  println("");

  parent = pathNdist[0];
  dst = pathNdist[1];

  for(i = 0 ; i < numvertices ; i++) {
    print("" + i + " : " + dst[i] + " " + i + "--" + parent[i]);
    println("");
  }
}

function mst(graph, V)
{
  dist = array(V);
  for(i = 0 ; i < V; i++) {
    dist[i] = 9999.9999;
  }

  parent = array(V);
  for(i = 0 ; i < V; i++) {
    parent[i] = -1;
  }

  visited = array(V);
  for(i = 0 ; i < V; i++) {
    visited[i] = false;
  }

  dist[0] = 0.0;
  parent[0] = 0;
  

  for(count = 0 ; count < V; count ++) {
    //Find min distance
    min = 999.999;
    index = -1;

    for(i = 0 ; i < V; i++) {
      if(dist[i] < min && false == visited[i]) {
        min = dist[i];
        index = i;
      }
    }

    if(-1 == index) {
      break;
    }
    //Finalize the current visited node
    visited[index] = true;

    //upate the ngbrs
    for(i = 0 ; i < V; i++) {
      if(i == index || 0.0 == graph[index][i]   
            || true == visited[i]) {
        continue;
      }

      if(graph[index][i] < dist[i]) {
        dist[i] = graph[index][i];
        parent[i] = index;
      }
    }
  }

  ret  = array(2);
  ret[0] = parent;
  ret[1] = dist;

  return ret;
}


function main(argv) 
{
  g = graph();
  g.loadFromFile("graphs/simple_mst.txt");
  //g.loadFromFile("graphs/02_graph_5_5_5.txt");
  //g.loadFromFile(argv[0]);

  vertices = g.getVertices();

  //matrix = g.getAdjacencyMatrix();

  //displayAdjacencyMatrix(matrix, vertices);


  mst_helper(g);
  
}



