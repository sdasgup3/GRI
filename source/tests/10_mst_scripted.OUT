   0 1 2 3 4 
0 |0 2 0 6 0 
1 |2 0 3 8 5 
2 |0 3 0 0 7 
3 |6 8 0 0 9 
4 |0 5 7 9 0 

0 : 0 0--0
1 : 2 1--0
2 : 3 2--1
3 : 6 3--0
4 : 5 4--1
